ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemAdd.java

```diff
@@ -594,13 +597,24 @@ private static void addClusteringServices(final OperationContext context, final
             ServiceInstaller installer = new ServiceInstaller() {
                 @Override
                 public ServiceController<?> install(RequirementServiceTarget target) {
-                    ServiceBuilder<?> builder = targetFactory.get().createSingletonServiceTarget(target).addService();
-                    return builder.setInstance(org.jboss.msc.Service.newInstance(builder.provides(CLUSTERED_SINGLETON_CAPABILITY.getCapabilityServiceName()), Boolean.TRUE))
-                            .setInitialMode(ServiceController.Mode.ON_DEMAND)
-                            .install();
+                    // Install on-demand singleton service
+                    // We don't want this to start until a deployment requires it
+                    ServiceController<?> controller = org.wildfly.service.ServiceInstaller.builder(Functions.constantSupplier(Boolean.TRUE))
+                            .provides(CLUSTERED_SINGLETON_CAPABILITY.getCapabilityServiceName())
+                            .build()
+                            .install(targetFactory.get().createSingletonServiceTarget(target));
+
+                    // Install well-known on-demand service that will force singleton service to start.
+                    ServiceInstaller.builder(Functions.constantSupplier(Boolean.TRUE))
+                            .provides(ServiceNameFactory.resolveServiceName(CLUSTERED_SINGLETON_BARRIER))
+                            .requires(controller.provides().stream().map(ServiceDependency::on).collect(Collectors.toList()))
```

Comment:
This could use a comment. Just reading the nearby code 'controller' provides one ServiceName but here there's complex stream processing happening.

Reply:
Added.